wgpu-core/src/device/resource.rs

```diff
@@ -3406,3 +3412,34 @@ impl<A: HalApi> Resource<DeviceId> for Device<A> {
         &mut self.info
     }
 }
+
+pub struct SnatchableMarker;
+use parking_lot::RwLockReadGuard;
+
+pub type SnatchGuard<'a> = RwLockReadGuard<'a, SnatchableMarker>;
+
+pub struct Snatchable<T> {
+    value: Option<T>,
+}
+
+impl<T> Snatchable<T> {
+    pub fn new(val: T) -> Self {
+        Snatchable { value: Some(val) }
+    }
+
+    pub fn get(&self, _guard: &SnatchGuard) -> Option<&T> {
```

Comment:
Note: To be perfectly transparent, one could argue that even passing the a lock guard, this method is not strictly safe because one could pass the wrong guard. Now we only have a single lock per device and there is no cross-device interaction allowed so one would really need to go out of their way to get this wrong.

I don't want to sprinkle `unsafe` over the entire code base for this but now is the time to argue if you feel otherwise or have a better solution.

Reply:
I think it's fine to have internal code not be fully up to unsafe standards, especially when the chance of getting it wrong is low. You could argue the majority of wgpu-core is unsafe, but that's would make the unsafe not really useful.